Instrument driver with OpenTelemetry tracing

wprzytula opened the pull request:
In order to increase driver's observability, it has been instrumented with OTel tracing. Requests, speculative executions and particular retries are covered with appropriate _spans_, which are filled with informative _tags_. The tracing incurs no overhead by default (when off), and bearable overhead when on. The PR contains (not many yet) tests and a Zipkin-based example.

wprzytula committed:
Tracing: introduced abstract tracing classes

In order to be generic and to still stick to Java 6 requirement for
driver-core, abstract classes: TracingInfo and TracingInfoFactory are
introduced. Further implementations of tracing are expected
to be placed apart from driver-core and hence keep Java 8 (or higher)
requirements away.

wprzytula committed:
Tracing: added customizable verbosity level

As some gathered data can be heavy or sensitive or we might want
to not have them in the trace for some reason, verbosity level layer
has been added. Currently, statement is both possibly long
and containing some sensitive data, so it is included only on level
FULL (and not NORMAL).

wprzytula committed:
Tracing: added abstract tracing classes to driver

While building a Cluster object, one can pass a TracingInfo class
implementation and it will be used for tracing in all sessions
with this cluster from now on.

wprzytula committed:
Tracing: added empty spans covering request

Every request (in RequestHandler) is covered by one "request" span,
and every (speculative) execution has its own "speculative_execution"
span. Each retry attempt is covered by "attempt" span, effectively
yielding a tree of form:
"request" -1--many-> "speculative_execution" -1--many-> "attempt"

wprzytula committed:
Tracing: added request status codes and exceptions

After a request is completed, its status code (OK or ERROR) is added
to trace data, as well as exception information (if one occured).

wprzytula committed:
Tracing: added abstract-based unit tests

These tests verify that the span tree structure is valid and that spans
are filled with tags of proper content, by providing a mock
implementation of TracingInfo.

wprzytula committed:
Tracing: added OpenTelemetry implementation

Added TracingInfo and TracingInfoFactory implementations in separate
driver-opentelemetry module.

wprzytula committed:
Tracing: added OpenTelemetry tracing example

Added OpenTelemetry and Zipkin connection configuration classes,
along with docker-compose file, example run script and related changes
in README.

wprzytula committed:
Tracing: added OpenTelemetry integration test

The test verifies that span tree structure and status code are valid.
Speculative executions are run parallel to the main thread, so some
of them can finish only after query result has been returned.
Thus, in order to collect span data from entire request, we decided
to wait until all speculative executions end. The main thread uses
conditional variable `allEnded` to wait for them and lock is used
for concurrent mutation of activeSpans.

avelanarius commented:
Some of the tests are failing. Those should be fixed by: https://github.com/zpp-2021-telemetry/java-driver/pull/10 - please look if this PR could be simplified (the reason it wasn't merged was that it was a bit too complex to understand and not documented well enough).

avelanarius commented:
Minor nit: formatting of commit messages: there shouldn't be a `.` at the end and line length of the body of the commit message shouldn't be longer than 72 characters (title no longer than 50 characters).

avelanarius commented on driver-opentelemetry/src/test/java/com/datastax/driver/opentelemetry/OpenTelemetryTest.java:
```diff
@@ -0,0 +1,171 @@
+              assertEquals(spanData.getName(), "request");
+              assertEquals(spanData.getStatus().getStatusCode(), StatusCode.OK);
+            });
+  }
```
Please add an additional test of a request that failed (can be an easy one as syntax error).

avelanarius commented on driver-opentelemetry/src/test/java/com/datastax/driver/opentelemetry/OpenTelemetryTest.java:
```diff
@@ -0,0 +1,171 @@
+        .map(ReadableSpan::toSpanData)
+        .forEach(
+            spanData -> {
+              assertEquals(spanData.getName(), "request");
```
Checking of other tags of span should be performed here (such as keyspace name).

avelanarius commented on driver-opentelemetry/src/test/java/com/datastax/driver/opentelemetry/OpenTelemetryTest.java:
```diff
@@ -0,0 +1,171 @@
+              Scope scope = userSpan.makeCurrent();
+
+              session.execute("INSERT INTO t(k, v) VALUES (4, 2)");
+              session.execute("INSERT INTO t(k, v) VALUES (2, 1)");
```
Other statement types should be tested (`UPDATE`, `SELECT`, `DELETE FROM`, `TRUNCATE TABLE`) whether they produce valid spans and resonable tags (row count not present in `TRUNCATE TABLE` for example).

wprzytula committed:
Tracing: added various request driver-side tags

All data that constitute added tags are already know by the driver,
i.e. they do not require fetching any additional metadata from
the cluster.

wprzytula commented:
After long adventures with failing tests, the PR is finally ready for last review.

wprzytula commented:
@avelanarius the requested fixes regading boundValues and partitionKey have been added.

wprzytula commented:
Renamed PrecisionLevel to VerbosityLevel.
